File: src/BarcodeReaderCode/VB.NET/BarcodeReader/BarcodeReader/Form1.vb
Imports System.Windows.Forms
Imports Dynamsoft.DotNet.TWAIN.Barcode
Imports Dynamsoft.DotNet.TWAIN.Enums.Barcode
Imports System.Text


Public Class Form1

    Private Sub Button1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnOpen.Click
        Dim filedlg As OpenFileDialog
        filedlg = New OpenFileDialog()
        filedlg.Multiselect = True
        ' try to locate images folder
        Dim imagesFolder As String = Application.ExecutablePath
        Dim pos As Integer = imagesFolder.LastIndexOf("\bin\")
        If (pos <> -1) Then
            imagesFolder = imagesFolder.Substring(0, imagesFolder.IndexOf(System.IO.Path.DirectorySeparatorChar, pos)) + "\Images\BarcodeImages\"
        Else
            pos = imagesFolder.LastIndexOf("\")
            imagesFolder = imagesFolder.Substring(0, imagesFolder.IndexOf("\", pos)) + "\"
        End If

        'use this folder as a starting point
        filedlg.InitialDirectory = imagesFolder

        Dim strFilename As String
        If (filedlg.ShowDialog() = DialogResult.OK) Then
            For Each strFilename In filedlg.FileNames

                DynamicDotNetTwain1.LoadImage(strFilename)

            Next
            dynamicDotNetTwain1_OnImageAreaDeselected(DynamicDotNetTwain1.CurrentImageIndexInBuffer)
        End If
    End Sub

    Private Sub Button2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnRecognize.Click
        Me.TextBox1.Text = ""
        Dim strDllFolder As String
        strDllFolder = Application.ExecutablePath
        Dim pos As Integer
        pos = strDllFolder.LastIndexOf("\bin\")
        If (pos <> -1) Then
            strDllFolder = strDllFolder.Substring(0, strDllFolder.IndexOf("\", pos)) + "\Redistributable\BarcodeResources\"
        Else
            pos = strDllFolder.LastIndexOf("\")
            strDllFolder = strDllFolder.Substring(0, strDllFolder.IndexOf("\", pos)) + "\"
        End If
        Me.DynamicDotNetTwain1.BarcodeDllPath = strDllFolder

        Try
            DynamicDotNetTwain1.MaxBarcodesToRead = Integer.Parse(tbxMaxNum.Text)
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Maximum Number Input error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Try

        If (Me.DynamicDotNetTwain1.CurrentImageIndexInBuffer < 0) Then
            MessageBox.Show("Please load an image before recognizing!", "Index out of bounds", MessageBoxButtons.OK, MessageBoxIcon.Warning)
            Return
        End If

        Me.TextBox1.Text = "Recognizing..."

        Dim aryResult() As Result

        Dim format As BarcodeFormat
        format = CType(System.Enum.Parse(GetType(BarcodeFormat), Val(cbxFormat.SelectedValue)), BarcodeFormat)

        aryResult = Me.DynamicDotNetTwain1.ReadBarcode(Me.DynamicDotNetTwain1.CurrentImageIndexInBuffer, Integer.Parse(tbxLeft.Text), Integer.Parse(tbxTop.Text), Integer.Parse(tbxRight.Text), Integer.Parse(tbxButtom.Text), format)
        Dim strText As StringBuilder
        strText = New StringBuilder()

        If aryResult.Length = 1 Then
            strText.AppendFormat(aryResult.Length & " total barcode" & ("") & " found." & Constants.vbCrLf)
        Else
            strText.AppendFormat(aryResult.Length & " total barcode" & ("s") & " found." & Constants.vbCrLf)
        End If

        For i As Integer = 0 To aryResult.Length - 1
            Dim objResult As Result
            objResult = aryResult(i)
            strText.AppendFormat("      Result " & (i + 1) & Constants.vbCrLf)
            strText.AppendFormat("      BarcodeFormat: " & objResult.BarcodeFormat.ToString() & Constants.vbCrLf)
            strText.AppendFormat("      Text read: " & objResult.Text & Constants.vbCrLf)
        Next i

        Me.TextBox1.Text = strText.ToString()
    End Sub

    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        InitializeUI()
    End Sub

    Protected Sub InitializeUI()
        cbxFormat.DataSource = [Enum].GetValues(GetType(BarcodeFormat))
        tbxMaxNum.Text = "10"
    End Sub

    Private Sub DynamicDotNetTwain1_OnImageAreaDeselected(ByVal sImageIndex As System.Int16) Handles DynamicDotNetTwain1.OnImageAreaDeselected
        tbxLeft.Text = "0"
        tbxTop.Text = "0"
        tbxRight.Text = DynamicDotNetTwain1.GetImage(DynamicDotNetTwain1.CurrentImageIndexInBuffer).Width.ToString()
        tbxButtom.Text = DynamicDotNetTwain1.GetImage(DynamicDotNetTwain1.CurrentImageIndexInBuffer).Height.ToString()

    End Sub

    Private Sub DynamicDotNetTwain1_OnTopImageInTheViewChanged(ByVal sImageIndex As System.Int16) Handles DynamicDotNetTwain1.OnTopImageInTheViewChanged
        Dim rect As Rectangle
        rect = DynamicDotNetTwain1.GetSelectionRect(DynamicDotNetTwain1.CurrentImageIndexInBuffer)
        If rect = Rectangle.Empty Then
            tbxLeft.Text = "0"
            tbxTop.Text = "0"
            tbxRight.Text = DynamicDotNetTwain1.GetImage(DynamicDotNetTwain1.CurrentImageIndexInBuffer).Width.ToString()
            tbxButtom.Text = DynamicDotNetTwain1.GetImage(DynamicDotNetTwain1.CurrentImageIndexInBuffer).Height.ToString()
        Else
            tbxLeft.Text = rect.Left.ToString()
            tbxTop.Text = rect.Top.ToString()
            tbxRight.Text = rect.Right.ToString()
            tbxButtom.Text = rect.Bottom.ToString()
        End If
    End Sub

    Private Sub DynamicDotNetTwain1_OnImageAreaSelected(ByVal sImageIndex As System.Int16, ByVal left As System.Int32, ByVal top As System.Int32, ByVal right As System.Int32, ByVal bottom As System.Int32) Handles DynamicDotNetTwain1.OnImageAreaSelected
        tbxLeft.Text = left.ToString()
        tbxTop.Text = top.ToString()
        tbxRight.Text = right.ToString()
        tbxButtom.Text = bottom.ToString()
    End Sub

    Private Sub DynamicDotNetTwain1_OnPostTransfer() Handles DynamicDotNetTwain1.OnPostTransfer
        DynamicDotNetTwain1_OnImageAreaDeselected(DynamicDotNetTwain1.CurrentImageIndexInBuffer)
    End Sub
End Class

File: src/BarcodeReaderCode/C#/BarcodeReader/BarcodeReader/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Dynamsoft.DotNet.TWAIN.Enums.Barcode;
using Dynamsoft.DotNet.TWAIN.Barcode;
using System.Collections;
using Dynamsoft.DotNet.TWAIN;

namespace BarcodeDemo
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
            InitializeUI();
        }

        protected void InitializeUI()
        {
            cbxFormat.DataSource = Enum.GetValues(typeof(Dynamsoft.DotNet.TWAIN.Enums.Barcode.BarcodeFormat));
            tbxMaxNum.Text = "10";
        }

        private void button1_Click(object sender, EventArgs e)
        {
            OpenFileDialog filedlg = new OpenFileDialog();
            filedlg.Multiselect = true;
            // try to locate images folder
            string imagesFolder = Application.ExecutablePath;

            // we assume we are running under the DotImage install folder
            imagesFolder = imagesFolder.Replace("/", "\\");
            int pos = imagesFolder.LastIndexOf("\\bin\\");
            if (pos != -1)
            {
                imagesFolder = imagesFolder.Substring(0, imagesFolder.IndexOf(@"\", pos)) + @"\Images\BarcodeImages\";
            }
            else
            {
                pos = imagesFolder.LastIndexOf("\\");
                imagesFolder = imagesFolder.Substring(0, imagesFolder.IndexOf(@"\", pos)) + @"\";
            }

            //use this folder as starting point			
            filedlg.InitialDirectory = imagesFolder;

            if (filedlg.ShowDialog() == DialogResult.OK)
            {
                foreach (string strfilename in filedlg.FileNames)
                {
                    this.dynamicDotNetTwain1.LoadImage(strfilename);
                }
                dynamicDotNetTwain1_OnImageAreaDeselected(dynamicDotNetTwain1.CurrentImageIndexInBuffer);
            }
        }

        private void button2_Click(object sender, EventArgs e)
        {
            this.textBox1.Text = "";
            string strDllFolder = Application.ExecutablePath;
            strDllFolder = strDllFolder.Replace("/", "\\");
            int pos = strDllFolder.LastIndexOf("\\bin\\");
            if (pos != -1)
            {
                strDllFolder = strDllFolder.Substring(0, strDllFolder.IndexOf(@"\", pos)) + @"\Redistributable\BarcodeResources\";
            }
            else
            {
                pos = strDllFolder.LastIndexOf("\\");
                strDllFolder = strDllFolder.Substring(0, strDllFolder.IndexOf(@"\", pos)) + @"\";
            }


            this.dynamicDotNetTwain1.BarcodeDllPath = strDllFolder;
            try
            {
                dynamicDotNetTwain1.MaxBarcodesToRead = int.Parse(tbxMaxNum.Text);
            }
            catch (Exception exp)
            {
                MessageBox.Show(exp.Message, "Maximum Number Input error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }

            if (this.dynamicDotNetTwain1.CurrentImageIndexInBuffer < 0)
            {
                MessageBox.Show("Please load an image before recognizing!", "Index out of bounds", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                return;
            }

            this.textBox1.Text = "Recognizing...";

            Result[] aryResult = this.dynamicDotNetTwain1.ReadBarcode(this.dynamicDotNetTwain1.CurrentImageIndexInBuffer, int.Parse(tbxLeft.Text), int.Parse(tbxTop.Text), 
                int.Parse(tbxRight.Text), int.Parse(tbxButtom.Text), (BarcodeFormat)cbxFormat.SelectedValue);   //int.Parse has not checked exception
            StringBuilder strText = new StringBuilder();
            strText.AppendFormat(aryResult.Length + " total barcode" + (aryResult.Length == 1 ? "" : "s") + " found.\r\n");
            for (int i = 0; i < aryResult.Length; i++)
            {
                Result objResult = aryResult[i];
                strText.AppendFormat("      Result " + (i + 1) + "\r\n");
                strText.AppendFormat("      BarcodeFormat: " + objResult.BarcodeFormat.ToString() + "\r\n");
                strText.AppendFormat("      Text read: " + objResult.Text + "\r\n");

            }
            this.textBox1.Text = strText.ToString();
        }

        private void dynamicDotNetTwain1_OnTopImageInTheViewChanged(short sImageIndex)
        {
            Rectangle rect = dynamicDotNetTwain1.GetSelectionRect(dynamicDotNetTwain1.CurrentImageIndexInBuffer);
            if (rect == Rectangle.Empty)
            {
                tbxLeft.Text = "0";
                tbxTop.Text = "0";
                tbxRight.Text = dynamicDotNetTwain1.GetImage(dynamicDotNetTwain1.CurrentImageIndexInBuffer).Width.ToString();
                tbxButtom.Text = dynamicDotNetTwain1.GetImage(dynamicDotNetTwain1.CurrentImageIndexInBuffer).Height.ToString();
            }
            else
            {
                tbxLeft.Text = rect.Left.ToString();
                tbxTop.Text = rect.Top.ToString();
                tbxRight.Text = rect.Right.ToString();
                tbxButtom.Text = rect.Bottom.ToString();
            }
        }

        private void dynamicDotNetTwain1_OnImageAreaSelected(short sImageIndex, int left, int top, int right, int bottom)
        {
            tbxLeft.Text = left.ToString();
            tbxTop.Text = top.ToString();
            tbxRight.Text = right.ToString();
            tbxButtom.Text = bottom.ToString();
        }

        private void dynamicDotNetTwain1_OnImageAreaDeselected(short sImageIndex)
        {
            tbxLeft.Text = "0";
            tbxTop.Text = "0";
            tbxRight.Text = dynamicDotNetTwain1.GetImage(dynamicDotNetTwain1.CurrentImageIndexInBuffer).Width.ToString();
            tbxButtom.Text = dynamicDotNetTwain1.GetImage(dynamicDotNetTwain1.CurrentImageIndexInBuffer).Height.ToString();
        }

        private void dynamicDotNetTwain1_OnPostAllTransfers()
        {
            dynamicDotNetTwain1_OnImageAreaDeselected(dynamicDotNetTwain1.CurrentImageIndexInBuffer);
        }

    }
}
